libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt

```diff
@@ -57,12 +57,13 @@ import okio.use
  */
 class WebSocketNetworkTransport
 private constructor(
-    private val serverUrl: String,
+    private val serverUrl: String?,
     private val headers: List<HttpHeader>,
     private val webSocketEngine: WebSocketEngine = DefaultWebSocketEngine(),
     private val idleTimeoutMillis: Long = 60_000,
     private val protocolFactory: WsProtocol.Factory = SubscriptionWsProtocol.Factory(),
     private val reopenWhen: (suspend (Throwable, attempt: Long) -> Boolean)?,
+    private val reopenServerUrl: (suspend () -> String)?,
```

Comment:
Nit: no need to add a second parameter there, you can only keep the suspend one and set it from the builder:

```kotlin
    fun build(): WebSocketNetworkTransport {
      if (serverUrl != null && reopenServerUrl != null) error("You can't have a static and a dynamic serverUrl at the same time")
      if (serverUrl == null && reopenServerUrl == null) error("No serverUrl specified")
      return WebSocketNetworkTransport(
          { serverUrl ?: reopenServerUrl!!.invoke() },
```

Reply:
Oh yes way better! I'll apply the changes :)